test/integration/quic_protocol_integration_test.cc

```diff
@@ -1,7 +1,59 @@
+#include "source/common/network/socket_option_impl.h"
+
 #include "test/integration/protocol_integration_test.h"
 
 namespace Envoy {
 
+// Test that the quiche code can handle packets getting batched together, i.e.
+// that it will re-register to read even without incoming packets.
+TEST_P(DownstreamProtocolIntegrationTest, BatchedPackets) {
+  if (downstreamProtocol() != Http::CodecType::HTTP3) {
+    return; // Testing H3 client talking to H3 upstream only.
+  }
+  setUpstreamProtocol(Http::CodecType::HTTP3);
+  initialize();
+
+  // Set up the transport factory so the codec client will have credentials to
+  // talk to the upstream.
+  quic_transport_socket_factory_ = IntegrationUtil::createQuicUpstreamTransportSocketFactory(
+      *api_, stats_store_, context_manager_, thread_local_, san_to_match_,
+      true /*connect to upstream*/);
+
+  // Connect directly to the upstream.
+  int upstream_port = fake_upstreams_[0]->localAddress()->ip()->port();
+  // Make sure the client receive buffer can handle all the packets without loss.
+  auto options = std::make_shared<Network::Socket::Options>();
+  options->emplace_back(std::make_shared<Network::SocketOptionImpl>(
+      envoy::config::core::v3::SocketOption::STATE_PREBIND,
+      ENVOY_MAKE_SOCKET_OPTION_NAME(SOL_SOCKET, SO_RCVBUF), 1024 * 100));
+
+  codec_client_ = makeHttpConnection(makeClientConnectionWithOptions(upstream_port, options));
+
+  // Send a request and a response that can not be handled in one read.
+  auto response = codec_client_->makeHeaderOnlyRequest(default_request_headers_);
+  waitForNextUpstreamRequest();
+  upstream_request_->encodeHeaders(default_response_headers_, false);
+  // Send more than the 32 packets that will be handled in one read pass.
```

Comment:
How to guarantee these 35 packets arrive as a batch in client connection's socket instead of in seperate event loops? Or is the expectation of this test that it shouldn't be flaky when they arrive in one batch?

Reply:
the best we can do is the block we have.  So we send them all, deadlock the backend, then do a read. We don't start doing the read until we confirm the backend is done with its send.  Also I did some backup logging to verify we got the reregister event just to be 100% sure we were sending enough packets in the common case.

Reply:
Can you comment that this is a best effort and because of the dead lock you added, more likely these packets will all arrive before the client resumes the event loop.

Reply:
Good idea. I had similar questions reading this.

Reply:
yeah I'm not sure if it's best effort or not - given loopback I suspect it's guaranteed but I think none of us know enough about kernel udp implementation to be sure so I won't claim that in comments.